chainstate/types/src/gen_block_index.rs

```diff
@@ -81,16 +81,43 @@ impl GenBlockIndex {
         }
     }
 
+    pub fn is_persistent(&self) -> bool {
+        match self {
+            GenBlockIndex::Block(b) => b.is_persistent(),
+            GenBlockIndex::Genesis(..) => true,
+        }
+    }
+
     pub fn chain_transaction_count(&self) -> u128 {
         match self {
             GenBlockIndex::Block(b) => b.chain_transaction_count(),
             GenBlockIndex::Genesis(_) => 0,
         }
     }
+
+    pub fn genesis(chain_config: &ChainConfig) -> Self {
+        Self::Genesis(Arc::clone(chain_config.genesis_block()))
+    }
 }
 
 impl From<BlockIndex> for GenBlockIndex {
     fn from(bi: BlockIndex) -> Self {
         GenBlockIndex::Block(bi)
     }
 }
+
+impl PartialEq for GenBlockIndex {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Self::Block(b1), Self::Block(b2)) => b1 == b2,
+            (Self::Genesis(g1), Self::Genesis(g2)) => {
+                let eq = Arc::ptr_eq(g1, g2);
+                debug_assert!(eq, "Attempt to compare different geneses");
+                eq
+            }
+            (Self::Block(_), Self::Genesis(_)) | (Self::Genesis(_), Self::Block(_)) => false,
+        }
+    }
+}
+
+impl Eq for GenBlockIndex {}
```

Comment:
Please remove these. Equality is not a proper thing for a block index.

Wherever you need this, replace it by comparing ids if necessary.